### ISSUE TITLE
DDP should set grad to None for (globally) unused params

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#28294]
Currently, DDP computes the param to bucket mapping upfront, and allreduce grads for all params in every iteration. For unused params, it will just set grad to zero. By doing this, optimizer cannot tell if a param indeed has a zero grad or it is not used in the current iteration. This could trigger convergence problems for optimizers with weight decay and momentum. 

However, DDP cannot simply set grad to None for local unused parameters, as local unused parameters might be used in other processes, and hence we still need to allreduce its grad. 

#### Proposed Solution

We could insert a header (either to every bucket or for the entire model) indicating whether a param is used or not locally in the current iteration. Then, we can allreduce/allgather this header and use that information to decide whether we should set grads to None.

cc @ezyang @gchanan @zou3519 @jerryzh168 @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528

**Comment by pritamdamania87** [comment]
> This could trigger convergence problems for optimizers with weight decay and momentum.

Could you elaborate a bit on this in terms of how this causes problems? Also, we do something similar in distributed autograd with tensors that are not used. I'm assuming a similar convergence problem might be seen there?

**Comment by mrshenli** [comment]
@pritamdamania87 

It depends on the optimizer behavior. The problem @baobablyh is trying to solve is that, say param X is not used in one iteration, then its weight should not decay. However, by setting grads to 0, the optimizer cannot distinguish whether the grad is indeed zero or the param is actually unused. So, if we can set grad to None for globally unused params, it will be an extra signal that the optimizer can make use of to fix the problem.

**Comment by baobablyh** [comment]
For example, SGD optimizer does weight decay and momentum if grad is zero however skipping for None grad:
https://github.com/pytorch/pytorch/blob/master/torch/optim/sgd.py#L89

**Comment by pietern** [comment]
@baobablyh Are you actively working on this? If not, we keep the "assigned" field empty (and this is why @soumith unassigned you from the issue). Let's discuss before we dive into a specific solution.

We know upfront which parameters will be used and which ones won't, through the unused parameter checking code. This means that before we start the backward pass, we can already build a bitmap to represent which parameters are unused. Then, if we compute a global bitwise AND between these maps, we know which parameters we can skip altogether (and keep None). What do you think?

**Comment by zhaojuanmao** [comment]
@baobablyh I'm assigning this back to you as @mrshenli mentioned you are actively working on this. 

The solution proposed by @pietern sounds good to me as well. 

@baobablyh which solution are you working on right now?

**Comment by baobablyh** [comment]
@zhaojuanmao I was working on the solution proposed by @mrshenli originally. I like the solution proposed by @pietern too. My concern is what the performance impact by the global bitwise would look like?

**Comment by zhaojuanmao** [comment]
my understanding is over head of both solutions is from all reduce, so the performance impact of both solutions should be similar. If you've already worked on Shen's solution and close to finish it, maybe send a PR you've already had first?

**Comment by mrshenli** [comment]
I agree with @zhaojuanmao. The overhead depends on your use case. 

* If your training loop uses a fixed set of params for say 1000 iterations, and then all switch to another set of params, I think @pietern's suggestion will result in a lighter overhead, as you would like to that bitwise check once every 1000 iteration. 
* If you training loop switch params very frequently, and if the unused params are a very small subset compared to the entire model, inserting that bitmap info into each bucket might be faster, as you don't need to wait for the bitwise allreduce to finish to determine the param to bucket mapping.
